src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,125 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROW_INDEX = 1.5;
+    private static final int START_INDEX_VALUE = 0;
+    private T[] dataArray;
+    private int size;
+
+    public ArrayList() {
+        dataArray = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    public ArrayList(int capacity) {
+        dataArray = (T[]) new Object[capacity];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == dataArray.length) {
+            dataArray = grow();
+        }
+        dataArray[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("index is invalid");
+        }
+        if (size == dataArray.length) {
+            dataArray = grow();
+        }
+        if (index == size) {
+            dataArray[size++] = value;
+        } else {
+            System.arraycopy(dataArray, index, dataArray, index + 1, size++);
+            dataArray[index] = value;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int oldSize = size;
+        int neededSize = size + list.size();
+        int indexOfValueFromList = START_INDEX_VALUE;
+        if (size == dataArray.length || dataArray.length < neededSize) {
```

Comment:
The condition `size == dataArray.length` is redundant here since `growToNeededSize` already checks if the `neededSize` is greater than the current array length. You can remove this condition.